crates/bevy_pbr/src/material_bind_groups.rs

```diff
@@ -1,145 +1,266 @@
+    buffer_dirty: BufferDirtyState,
+    phantom: PhantomData<M>,
+}
 
-    /// A bitfield that contains a 0 if the slot is free or a 1 if the slot is
-    /// full.
+/// A single binding array for storing bindless resources and the bookkeeping
+/// necessary to allocate into it.
+struct MaterialBindlessBindingArray<R>
+where
+    R: GetBindingResourceId,
+{
+    /// The number of the binding that we attach this binding array to.
+    binding_number: BindingNumber,
+    /// A mapping from bindless slot index to the resource stored in that slot,
+    /// if any.
+    bindings: Vec<Option<MaterialBindlessBinding<R>>>,
+    /// The type of resource stored in this binding array.
+    resource_type: BindlessResourceType,
+    /// Maps a resource ID to the slot in which it's stored.
     ///
-    /// We keep this value so that we can quickly find the next free slot when
-    /// we go to allocate.
-    used_slot_bitmap: u32,
+    /// This is essentially the inverse mapping of [`Self::bindings`].
+    resource_to_slot: HashMap<BindingResourceId, u32>,
+    /// A free list of slots in [`Self::bindings`] that contain no binding.
+    free_slots: Vec<u32>,
+    /// The number of allocated objects in this binding array.
+    len: u32,
 }
 
-/// Information that the allocator keeps about each bind group for which
-/// bindless textures are not in use.
+/// A single resource (sampler, texture, or buffer) in a binding array.
 ///
-/// When a bindless texture isn't in use, bind groups and material instances are
-/// in 1:1 correspondence, and therefore there's only a single slot for extra
-/// material data here.
-pub struct MaterialNonBindlessBindGroup<M>
+/// Resources hold a reference count, which specifies the number of materials
+/// currently allocated within the slab that refer to this resource. When the
+/// reference count drops to zero, the resource is freed.
+struct MaterialBindlessBinding<R>
+where
+    R: GetBindingResourceId,
+{
+    /// The sampler, texture, or buffer.
+    resource: R,
+    /// The number of materials currently allocated within the containing slab
+    /// that use this resource.
+    ref_count: u32,
+}
+
+/// The allocator that stores bind groups for non-bindless materials.
+pub struct MaterialBindGroupNonBindlessAllocator<M>
 where
     M: Material,
 {
-    /// The single allocation in a non-bindless bind group.
-    allocation: MaterialNonBindlessBindGroupAllocation<M>,
+    /// A mapping from [`MaterialBindGroupIndex`] to the bind group allocated in
+    /// each slot.
+    bind_groups: Vec<Option<MaterialNonBindlessAllocatedBindGroup<M>>>,
+    /// The bind groups that are dirty and need to be prepared.
+    ///
+    /// To prepare the bind groups, call
+    /// [`MaterialBindGroupAllocator::prepare_bind_groups`].
+    to_prepare: HashSet<MaterialBindGroupIndex>,
+    /// A free list of bind group indices.
+    free_list: Vec<MaterialBindGroupIndex>,
```

Comment:
free lists are a specific kind of data structure https://en.wikipedia.org/wiki/Free_list
```suggestion
    /// A list of free bind group indices.
    free_indices: Vec<MaterialBindGroupIndex>,
```